keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,962 @@
+and make progress.
+-->
+- Support richer policies in addition to `RequiredNotPreempted` and `PreferNotPreempted`
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I deployed a service in a cluster and to ensure its high availability, 
+I created PDB for this deployment and set '.spec.minAvailable' to 3. I want the PDB to be 
+guaranteed even in case of scheduler preemption.
+
+#### Story 2
+I created a Tensorflow distributed job that requires a minimum of 5 workers running.
+I created PDB for the job and set '.spec.minAvailable' to 5. I want the PDB to be 
+guaranteed even in case of scheduler preemption to ensure the stability of the whole job.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+If a user sets a PodDisruptionBudget (PDB) for some low-priority pods and sets the `PreemptionPolicy` 
+to `RequiredNotPreempted`, high-priority pods will not be able to violate the PDB and preempt 
+these pods during the scheduling process. This may result in high-priority pods being unable 
+to schedule while low-priority pods continue to run normally, which may make troubleshooting 
+more complex for the user. To address this, additional logging or event descriptions will be 
+added in the implementation to clearly inform the user of the reason why preemption did not occur.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A new field named `PreemptionPolicy` will be added to `PodDisruptionBudgetSpec`.
```

Comment:
Are PDB authors the right people to make this decision, or is the PriorityClass author the right person? It seems wrong to let a namespace-constrained user who is only allowed to create low priority pods to consume cluster resources and block high priority pods from preempting.

At first glance, I would have expected preemption behavior to be specified in PriorityClass, as a "pods with this priority class should only be preempted when allowed by PDB" option

Reply:
> It seems wrong to let a namespace-constrained user who is only allowed to create low priority pods to consume cluster resources and block high priority pods from preempting.

Thanks. Your comments is greatly appreciated. We had also considered adding a field in the PriorityClass to indicate a behavior where preemption isn't allowed with violating PDB in the draft of our previous design document. https://docs.google.com/document/d/1g2wj8S04M0yTy2t4l4GzrdARtxBpkBTfkJHw0oqBvXM/edit#

However, as you have mentioned, the potential conflict between the creators of PDB and PriorityClass, who may have different priorities(cluster scope and namespace scope), has been overlooked and may result in high priority pods being blocked from preemption in the cluster. 

So, based on the above issues and your suggestions, we can address this by adding a field in the PriorityClass, which would allow for this behavior to be defined and managed consistently by the cluster administrator in the PriorityClass, rather than relying on the creators of PDB and PriorityClass potentially having conflicting priorities. 

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy

	//  PreemptedPolicy describes a policy for determining if the pod can be preempted
        //  by other higher priority pod.
	PreemptedPolicy *core.PreemptedPolicy
}

//  PreemptedPolicy describes a policy for determining if the pod can be preempted
//  by other higher priority pod.
type PreemptedPolicy string

const (
	// PreemptedAllowedByPDB indicates that a pod will not be preempted by higher
	// priority pods if it would result in a violation of the PDB.
	PreemptedAllowedByPDB PreemptedPolicy = "PreemptedAllowedByPDB"
	// PreemptedAlways indicates that a pod will be preempted by higher priority
	// pods at any time.
	PreemptedAlways PreemptedPolicy = "PreemptedAlways"
)
```
WDYT? @liggitt @alculquicondor

Reply:
That looks better, but I don't like the name.

Maybe PreemptionTargetPolicy?

Reply:
Thanks. In terms of naming, I do not have a clear preference as long as it can accurately expresses the policy for selecting a pod as a victim when it is preempted.

Reply:
@liggitt  Could you please help to take a look if it looks good to you? If so, I will modify the design document according to this idea? Thanks.

Reply:
Placing the field on priorityClass makes sense, though the name is still confusing to me. Would something like `DisruptionPolicy` be more clear?

```go
// DisruptionPolicy specifies how pods with this priority class may be disrupted when being preempted.
DisruptionPolicy *DisruptionPolicy
```

for possible values, it seems like the current behavior is not quite "Always", it *tries* to honor PDB in a best-effort / if-possible sort of way. Would names that reflect that be clearer?
```
// - PodDisruptionBudgetBestEffort
// - PodDisruptionBudgetStrict
```

Reply:
+1 for moving the responsibilities to the priorityClass. Does not seems right for application owners affect priorities that they do not own.


It might be preferable to mention the preemption in the field name, as there might be other reason for disruption than preemption that we might want to configure per priority class in the future. Or maybe just name it `PodDisruptionBudgetPolicy` to be completely clear?